Javascript backend - run node package install once per workspace

riisi opened the pull request:
Fixes #19236 and fixes #19235.

I've been poking around at this backend for a while trying to understand it. I think I'm most of the way there...
I wanted to raise this PR now for feedback - I think it's fairly complete as is - but I'm not sure if I have any blind spots in both understanding the original design intent and how it's being used.

There is a lot more functionality than I expected going in - huge props to @tobni for that -  it would be great to fix the remaining issues and get some docs out so it can be used more & we can get more feedback.

## Current process

For each first party package, the backend runs 'npm install' on its first party dependency packages. This is a recursive process.

Each `npm install` command is executed in the context of that package, generating a 'node_modules' directory containing 1st and 3rd party deps (in separate directories). Source code (from `javascript_sources` targets) is then copied in. 
The output of `node_modules` from any 1st party dependency installs is merged to form `node_modules` for the package.

### Problems

Per #19235, newer versions of package managers generate integrity files in each node_modules directories. These cause conflicts on merge.

Secondly, and more critically, it seems (I've not dug too much into this, but some past discussion [here](https://pantsbuild.slack.com/archives/C0D7TNJHL/p1685026807993119)), that each `npm install` is effectively installing all transient deps.

E.g. in the case of a tree like package-A -> package-B -> package-C -> package-D ... I think while the intent is for package-A install to only install direct dependents, it's installing all transitive, so there's a lot more IO than intended. Effectively, it is merging ((package-A + trans deps) + (package-B + trans deps) + (package-c + trans deps)...))

## New approach

Perform one install for the package in the workspace context. 
The install command is `npm install --workspace package-a`. 

The install command needs to have visibility of the root package.json, which contains the 'workspaces' configuration, as well as any dependent package.json. Currently this is all package.json in the project. 

The output is a single node_modules directory at the project level (assuming default package manager config - hoisting options could be configured differently). 

### Advantages

Simpler and is similar to how the user would use the package manager natively to install in a mono-repo setup. 
The backend is simpler to troubleshoot as there is only one sandbox per install.
We delegate more of the package resolution process to the package managers.
Performance should be theoretically better given the issue above - but I've not tested. If anyone has anything handy for testing, please let me know. 

### Disadvantages

We are not caching each module install within Pants, which I'm guessing was the original intent. But I'm not sure we're gaining much from this approach currently.

tobni reviewed (commented):
Thanks! Happy to see contributions to the JS backend.

1. There's some unused fields / code that should be removed,
2. There ought to be at least one test case that support the implementation. I suspect all tests pass because of how much heavy lifting the package managers already do. At least exposing the integrity file case as passing would be a good case to cover.
3. Did you consider https://github.com/pantsbuild/pants/issues/19236? I'm fine with this change as is because it is a path forward, but I'm curious if you spent any thought on it.

tobni commented on src/python/pants/backend/javascript/README.md:
```diff
@@ -0,0 +1,140 @@
+This README is mainly for understanding the current backend functionality for PR review purposes. 
+Plan to migrate much of this content to user docs.
+
+## Definitions
```
I only skimmed through this README, LMK if you want it thoroughly checked.

huonw commented:
> Fixes https://github.com/pantsbuild/pants/issues/19236 and https://github.com/pantsbuild/pants/issues/19235.

(Totally minor, but FYI GitHub's auto-closing doesn't handle "and" like this. I've adjusted it to repeat the `fixes`. 😄 )

riisi commented:
Thanks for reviewing so quickly! It will probably take me a while to go back through everything. 

> 3. Did you consider [Install entire node project when creating node_modules digest, subset in sandboxes #19236](https://github.com/pantsbuild/pants/issues/19236)? I'm fine with this change as is because it is a path forward, but I'm curious if you spent any thought on it.

Only briefly. I've just been trying to find a path forward mainly while trying to understand the backend at the same time. 

If I'm understanding that correctly, it's suggesting to, instead of running `npm install --workspace package-a` for each workspace, run `npm install --workspaces` (i.e. install workspaces for the whole project at once) and then extract the results for each workspace? 

It seems to me there would be a couple of disadvantages:
- if a user is only working with a particular workspace, the whole project needs to be installed - if there are issues affecting other parts of the project, (e.g. a. missing package, or package conflicts, as hypotheticals), I was thinking it might be better to be able to only need to install a single workspace
- If the lockfile is updated (or other upstream cache invalidation), then all workspaces need to be re-installed which would take longer - unless there's a way around this?

riisi commented on src/python/pants/backend/javascript/nodejs_project_test.py:
```diff
@@ -74,7 +74,10 @@ def test_parses_projects(rule_runner: RuleRunner) -> None:
     ("package_manager", "expected_immutable_install_args"),
     [
-        (None, ("clean-install",)),
+        (
```
Something is different with the version of black I have I guess. Reverted it.

riisi commented on src/python/pants/backend/javascript/nodejs_project.py:
```diff
@@ -222,10 +222,14 @@ async def find_node_js_projects(
+            ProjectPaths(pkg.root_dir, ["", *pkg.workspaces])
+            if pkg not in pnpm_workspaces
+            else ProjectPaths(pkg.root_dir, ["", *pnpm_workspaces[pkg].packages])
+        )
```
Reverted

riisi commented on src/python/pants/backend/javascript/subsystems/package-lock.json:
```diff
@@ -1,794 +1,464 @@
-            }
-        }
+  "name": "@the-company/project",
+  "lockfileVersion": 3,
```
Think I may have accidentally modified this - reverted it.

riisi commented on src/python/pants/backend/javascript/README.md:
```diff
@@ -0,0 +1,140 @@
+
+No import, package.json dep:
+- node_modules missing source code for 1st party deps, everything else present
+- Probably in this case, it would be nice to generate a warning, e.g. "Package-a source code has a dependency on package-b but package-b is not present in package-a's listed dependencies in package.json".
```
That's true - that would work.

riisi commented:
I've reworked the package integration test to cover the case of importing another workspace package, which triggered the integrity file conflict. In doing so, I found a bug, which held me up for a while. 

A newer version of pnpm (>8) than the default for node 16 (6.11.0) is needed for workspaces support. I found that the version specified on the subsystem options was not being used. Although it was being installed in a separate sandbox, the output was not captured, and on the process invocation for `pnpm install`, corepack was installing the default version. 

I've left a couple of comments on the fix for now - I'm just not sure if I've missed anything re. the original intent for the `prepare_corepack_tool` rule @tobni ?

riisi commented on src/python/pants/backend/javascript/install_node_package.py:
```diff
@@ -91,39 +90,36 @@ async def install_node_packages_for_address(
         [PackageJsonSourceField], transitive_tgts.dependencies, union_membership
     )
     assert target not in pkg_tgts
```
Removed it

tobni commented on src/python/pants/backend/javascript/subsystems/nodejs.py:
```diff
@@ -556,12 +556,16 @@ async def prepare_corepack_tool(
+            # input_digest=request.input_digest,
             immutable_input_digests=environment.immutable_digest(),
-            working_directory=request.working_directory,
+            # The working directory should be the build root - if not, the output_directory will not be collected
+            # working_directory=request.working_directory,
```
The reason for this being true should be investigated. There should not be a requirement for this tool to be ran using the builld root as cwd in practice, and pants would not want to enforce that a `package.json` file is at this position.

I suspect the `environment.corepack_home` variable is a relative path instead of an absolute one, as you noted in the comment below.

tobni replied:
To be clear, the reason we mount the package.json and move to its containing folder as `CWD` is to trigger corepacks default behaviour of reading the `package.json` of where it was invoked, and then have corepack persist the result in the location of `environment.corepack_home`

tobni replied:
Essentially `corepack_home` is missing a `{chroot}` token.

tobni replied:
https://github.com/pantsbuild/pants/pull/21021 Incorporpated your observation in the bugfix.

tobni commented on src/python/pants/backend/javascript/subsystems/nodejs.py:
```diff
@@ -556,12 +556,16 @@ async def prepare_corepack_tool(
             ),
             description=f"Preparing configured {tool_description}.",
-            input_digest=request.input_digest,
+            # TODO: Given below, is there ever any need to supply an input_digest (e.g. package.json files) here?
+            # as far as I can see, the version is always resolved for the workspace and passed in as a CLI arg to corepack
+            # input_digest=request.input_digest,
```
Yes, I believe supporting the default style of specifying package manager versions in `package.json` has real value. I think the issue is I accidentally only tested this with a `package.json` that specifies a pkg manager requirement at the build root, but it shouldn't be a requirement (as hinted by the passing of a working directory).

tobni reviewed (commented):
While the new bug you've found is real, I would argue keeping it to a separate issue and pull request is cleaner. I can look into this.

To make your integration test pass I believe removing `src/js` from the path will work as a temporary.

w.r.t to the pnpm version, I think maybe best we can do is bump the default? It is unfortunate, but I'm not keen on validating tool versions for features. I believe pants should aim to please instead of throwing errors that might not be relevant for a particular usecase. Users can use the `nodejs` subsystem and its package managers without writing javascript, after all.

riisi commented:
@tobni I've taken a stab at updating the node version and removing `/src/js` from the integration test files. 
This does seem to resolve the bug above, as well as providing a pnpm version that supports workspaces. 

However, I still have one more problem when the test is run without explicitly providing package manager versions (e.g. `"--nodejs-package-managers={'npm': '7.20.1', 'pnpm': '8.15.5', 'yarn': '1.22.15'}",`) - I get the error `The local project doesn't feature a 'packageManager' field - please explicit the package manager to pack, or update the manifest to reference it`. 

In the file layout scenario used in the test, the workspace root shouldn't contain a package.json (for pnpm and yarn), and not necessarily a package manager. 

To address this for now, I modified the logic in prepare_corepack_tool() to supply the package manager name, but seems undesirable as this would override the package.json spec if it was provided. 

_However_, in my testing it seems that if a version is provided in package.json, then it is used: i.e.

```js
@rule(desc="Preparing Corepack managed tool.")
async def prepare_corepack_tool(
    request: CorepackToolRequest, environment: NodeJSProcessEnvironment, nodejs: NodeJS
) -> CorepackToolDigest:
    version = request.version or nodejs.package_managers.get(request.tool) # <-- version here is that from package.json and fed into CLI args
    if not version and request.input_digest == EMPTY_DIGEST:
        raise ValueError(f"Could not determine tool version for {request.tool}.")
    tool_spec = f"{request.tool}@{version}" if version else request.tool
    tool_description = tool_spec if version else f"default {tool_spec} version"
```
so this doesn't actually matter. 

This goes back to my earlier comment above on the same section of code:
```
            # TODO: Given below, is there ever any need to supply an input_digest (e.g. package.json files) here?
```

Thoughts on what to do here? 


Aside from this, there are some failing tests, I think all related to package-lock files needing updating after the node updates. 
I can fix - just want to avoid rework in case we need to backtrack on anything above.

tobni commented:
> Essentially `corepack_home` is missing a `{chroot}` token.

Did you try this? To clarify, the environment variable COREPACK_HOME is not correctly set when preparing the tool. This is why the `working_directory` setting is not working, and why removing 1 level of folders:
> I've taken a stab at updating the node version and removing /src/js from the integration test files.

works. `{chroot}` should be added to the variable in the sandbox so that it is an absolute path. It is currently accidentally relative.

>However, I still have one more problem when the test is run without explicitly providing package manager versions (e.g. "--nodejs-package-managers={'npm': '7.20.1', 'pnpm': '8.15.5', 'yarn': '1.22.15'}",) - I get the error The local project doesn't feature a 'packageManager' field - please explicit the package manager to pack, or update the manifest to reference it.

Had a little look through and here's what I gather:

I was fairly certain that corepack vendored their own releases of package managers so that e.g `corepack prepare pnpm --activate` worked OOTB if no other version was provided. Testing now reveals I was mistaken, in order for that mode of operation to work the required invocation would be `corepack prepare --all --activate`. Unfortunately this invocation downloads all package managers, but I think that price tag is worth it.

I think that would be the desired behaviour here. I.e the rules _should be_

1. If `package.json#packageManager` contains a package manager version, it should be used; otherwise
2. If `pants.toml#[nodejs].package_managers` specifies a version for a package manager, this should be used; otherwise
3. Lastly we should fall back on the version corepack provides via its "last known good releases".

I think what is _actually_ missing is case 3, which would be different from the wip you've pushed to remote.

And I agree, as you've spotted, there is no reason to provide a digest because the relevant bit (if present) to achieve 1 is already parsed earlier, so there's no need for corepack to do it again! I suspect the addition of `install_from_resolve` support is what caused the digest to become unnecessary.

Lastly:
I think a nodejs version upgrade has to be a separate pull request from the rest of these changes. It needs to be **very clear** in release documentation that the default nodejs version is being bumped.

tobni commented:
I'm going to address the corepack issues via #21020 to cut out the noise from the changes herein.

riisi commented:
Thanks! Sorry have been flat out this week  On 6 Jun 2024, at 20:08, Tobias Nilsson ***@***.***> wrote:﻿
I'm going to address the corepack issues via #21020 to cut out the noise from the changes herein.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

riisi commented:
I removed the last WIP commit with the node upgrade and rebased after your fix was merged @tobni. It works fine now in a subdir. 

The update to pnpm v8 is causing a coverage test to fail - see comment [here](https://github.com/pantsbuild/pants/pull/21012#issuecomment-2183902587). I think we should do the node upgrade first and fix that there.

tobni commented:
@riisi Headsup that nodejs18 PR was just merged :)

riisi commented:
@tobni Updated this now following @krishnan-chandra 's changes.

tobni commented on src/python/pants/backend/javascript/nodejs_project.py:
```diff
@@ -236,6 +236,8 @@ async def find_node_js_projects(
     nodejs: NodeJS,
     resolve_names: UserChosenNodeJSResolveAliases,
 ) -> AllNodeJSProjects:
+    # Note: If pnpm_workspace.yaml is present for an npm-managed project, it will override the package.json["workspaces"] setting, which is not intuitive
```
I took the liberty of creating https://github.com/pantsbuild/pants/issues/21134, please link it in this comment.

riisi replied:
Added it now.

tobni commented on src/python/pants/backend/javascript/README.md:
```diff
@@ -0,0 +1,89 @@
+  - area-A/
+    - package.json with 'workspaces' defined (yarn, npm)
+      - workspaces: ["packages/package-a", "packages/package-b", "packages/sub-dir/package-c"]
+    - \[pnpm-workspace.yaml\] - pnpm only - will cause issues if present for npm/yarn
```
Same here, link the issue. This shouldn't be a hard problem to fix. https://github.com/pantsbuild/pants/issues/21134

riisi replied:
I removed the README as I figured it would add a lot of noise to this PR at this stage, and it was mainly for me to work through things and clarify assumptions.  Prefer to use it as input to documentation instead, but can add here or separately if you think it's helpful for others.

tobni commented on src/python/pants/backend/javascript/nodejs_project_environment.py:
```diff
@@ -125,12 +125,6 @@ async def setup_nodejs_project_environment_process(req: NodeJsProjectEnvironment
         args = tuple(req.args)
     output_files = req.output_files
     output_directories = req.output_directories
```
Why is the per_package_caches removed?

riisi replied:
I think I may have been thinking it was no longer necessary but I'm not sure why I thought that, and the removal here hardly seems comprehensive. Or maybe it was when trying to track down the bug related to corepack home. 
Anyway, put it back.

tobni reviewed (approved):
Awesome, thanks!
